refactor(linter): add fixer for `typescript-eslint/no-non-null-asserted-optional-chain`

camchenry opened the pull request:
Adds fixes for the `typescript-eslint/no-non-null-asserted-optional-chain` rule in the linter and adds better diagnostic messages with labeled spans.

camchenry commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9993?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#9993** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9993?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9993?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#9991** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9991?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=9993).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

shulaoda commented on crates/oxc_language_server/src/linter/snapshots/fixtures_linter_issue_9958_issue.ts.snap:
```diff
@@ -15,14 +15,26 @@ tags: None
 tags: None
+            
+
+code: "None"
+code_description.href: "None"
+message: "non-null assertion made after optional chain"
+range: Range { start: Position { line: 11, character: 21 }, end: Position { line: 11, character: 22 } }
```
Why did this change introduce the new snapshot?

camchenry replied:
It's related to the previous PR: when a diagnostic has multiple spans associated, we don't produce additional diagnostics unless the message content is non-empty. VS Code integration seemed to not work when the message content was empty. But now that I've added labels for these spans, we are able to report these diagnostics again, much like we do in the CLI.

camchenry commented:
### Merge activity

* **Mar 23, 10:19 PM EDT**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc) once it meets the requirements.

camchenry committed:
refactor(linter): add fixer for `typescript-eslint/no-non-null-asserted-optional-chain` (#9993)

Adds fixes for the `typescript-eslint/no-non-null-asserted-optional-chain` rule in the linter and adds better diagnostic messages with labeled spans.
